File: JavaPartI/la_stringAbbreviationMatching.java
// recursion
public class Solution {
  public boolean match(String input, String pattern) {
    // Write your solution here.
    return matchHelper(input, pattern, 0, 0);
  }
  private boolean matchHelper(String s, String t, int si, int ti) {
    if (si == s.length() && ti == t.length()) {
      return true;
    } 
    if (si >= s.length() || ti >= t.length()) {
      return false;
    }
    if (t.charAt(ti) < '0' || t.charAt(ti) > '9') {
      if (s.charAt(si) == t.charAt(ti)) {
        return matchHelper(s, t, si + 1, ti + 1);
      } 
      return false;
    } else {
      int count = 0;
      while (ti < t.length() && t.charAt(ti) >= '0' && t.charAt(ti) <= '9') {
        count = count * 10 + (t.charAt(ti) - '0');
        ti++;
      }
      return matchHelper(s, t, si + count, ti);
    }
  }
}

//// iteration: firstly check alphabetic, then digit, consider length prolblems of two strings
public class Solution2 {
  public boolean match(String input, String pattern) {
    // Write your solution here.
    if (input.length() == 0 && pattern.length() == 0) {
      return true;
    } else if (input.length() == 0 || pattern.length() == 0) {
      return false;
    } 
    int j = 0, i = 0;
    while (j < pattern.length()) {
      if (i >= input.length()) {
        return false;
      }
      if ((pattern.charAt(j) - '0') < 0 || (pattern.charAt(j) - '0') > 9) { // it is an alphabatic 
        if (input.charAt(i) != pattern.charAt(j)) {
          return false;
        }
        i++;
        j++;
      } else {  // it is a digit
        int prevsum = 0;
        while (j < pattern.length() && (pattern.charAt(j) - '0') >= 0 && (pattern.charAt(j) - '0') <= 9) {
            prevsum = prevsum * 10 + (pattern.charAt(j) - '0');
            j++;
        }
        i += prevsum;
      }
    }
    return i == input.length();

  }
}

File: pythonPart/la_stringAbbreviationMatching.py
class Solution(object):
  def match(self, input, pattern):
    """
    input: string input, string pattern
    return: boolean
    """
    # write your solution here
    
    n1 = len(input)
    n2 = len(pattern)
    if n1 == 0 and n2 == 0:
      return True
    if n1 == 0 or n2 == 0 or n1 < n2:
      return False
    i , j = 0, 0
    while i < n1 and j < n2:
      ch = pattern[j]
      if ch.isdigit():
        num = 0
        while j < n2 and pattern[j].isdigit():
          num = num * 10 + ord(pattern[j]) - ord('0')
          j += 1
       	while num > 0:
          i += 1
          num -= 1
      else:
        if (i < n1 and j < n2 and input[i] != pattern[j]):
          return False
        i = i + 1
        j = j + 1         
    return i == n1 and j == n2
